src/avram/save_operation.cr

```diff
@@ -339,6 +339,37 @@ abstract class Avram::SaveOperation(T)
     !!record_id
   end
 
+  # `#persisted?` always returns `true` in `after_*` hooks, whether
+  # a new record was created, or an existing one was updated.
+  #
+  # This method should always return `true` for a create or `false`
+  # for an update, independent of the stage we are at in the operation.
+  def new_record? : Bool
+    {{ T.resolve.constant(:PRIMARY_KEY_NAME).id }}.value.nil?
+  end
+
+  def revert : self?
+    return unless saved?
+
+    saved_record = record.not_nil!
+    operation = self.class.new(saved_record)
```

Comment:
If the SaveOperation has specific `needs`, then the new operation won't get them. When it calls `save` below, it'll run all of the callbacks which may end up failing.

Reply:
Good catch! Totally missed that part. Will see what I can come up with.